components/brave_wallet/browser/zcash/zcash_create_orchard_to_orchard_transaction_task.cc

```diff
@@ -0,0 +1,152 @@
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_create_orchard_to_orchard_transaction_task.h"
+
+#include <utility>
+
+#include "base/numerics/checked_math.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "components/grit/brave_components_strings.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace brave_wallet {
+
+ZCashCreateOrchardToOrchardTransactionTask::
+    ZCashCreateOrchardToOrchardTransactionTask(
+        absl::variant<
+            base::PassKey<class ZCashCreateOrchardToOrchardTransactionTaskTest>,
+            base::PassKey<ZCashWalletService>> pass_key,
+        ZCashWalletService& zcash_wallet_service,
+        ZCashActionContext context,
+        const OrchardAddrRawPart& receiver,
+        std::optional<OrchardMemo> memo,
+        uint64_t amount,
+        ZCashWalletService::CreateTransactionCallback callback)
+    : zcash_wallet_service_(zcash_wallet_service),
+      context_(std::move(context)),
+      receiver_(receiver),
+      memo_(memo),
+      amount_(amount),
+      callback_(std::move(callback)) {}
+
+ZCashCreateOrchardToOrchardTransactionTask::
+    ~ZCashCreateOrchardToOrchardTransactionTask() = default;
+
+void ZCashCreateOrchardToOrchardTransactionTask::Start() {
+  DCHECK(!started_);
+  started_ = true;
+  ScheduleWorkOnTask();
+}
+
+void ZCashCreateOrchardToOrchardTransactionTask::ScheduleWorkOnTask() {
+  base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+      FROM_HERE,
+      base::BindOnce(&ZCashCreateOrchardToOrchardTransactionTask::WorkOnTask,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashCreateOrchardToOrchardTransactionTask::WorkOnTask() {
+  if (error_) {
+    std::move(callback_).Run(base::unexpected(*error_));
+    zcash_wallet_service_->CreateTransactionTaskDone(this);
+    return;
+  }
+
+  if (!spendable_notes_) {
+    GetSpendableNotes();
+    return;
+  }
+
+  if (!transaction_) {
+    CreateTransaction();
+    return;
+  }
+
+  std::move(callback_).Run(base::ok(std::move(*transaction_)));
+  zcash_wallet_service_->CreateTransactionTaskDone(this);
+}
+
+void ZCashCreateOrchardToOrchardTransactionTask::GetSpendableNotes() {
+  context_.sync_state->AsyncCall(&OrchardSyncState::GetSpendableNotes)
+      .WithArgs(context_.account_id.Clone())
+      .Then(base::BindOnce(
+          &ZCashCreateOrchardToOrchardTransactionTask::OnGetSpendableNotes,
+          weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashCreateOrchardToOrchardTransactionTask::OnGetSpendableNotes(
+    base::expected<std::vector<OrchardNote>, OrchardStorage::Error> result) {
+  if (!result.has_value()) {
+    error_ = result.error().message;
+    ScheduleWorkOnTask();
+    return;
+  }
+
+  spendable_notes_ = result.value();
+  ScheduleWorkOnTask();
+}
+
+void ZCashCreateOrchardToOrchardTransactionTask::CreateTransaction() {
+  CHECK(spendable_notes_);
```

Comment:
Just spotted this instance as well of `CHECK`. Should this be changed to an if error catch or DCHECK to avoid a crash here?

Reply:
CHECK is fine here, since there will be a crash anyway when spendable_notes_ is accessed. And actually spendable_notes_ error case is handled in the OnGetSpendableNotes method.